File: server_python/controllers/admin_controller.py
import os
import jwt
from datetime import datetime
from bson import ObjectId
from configs.db import get_db
from utils.serialize import serialize_blog, serialize_comment

async def admin_login(email: str, password: str):
    if email != os.getenv("ADMIN_EMAIL") or password != os.getenv("ADMIN_PASSWORD"):
        return {"success": False, "message": "Invalid Credentials"}
    token = jwt.encode({"email": email}, os.getenv("JWT_SECRET"), algorithm="HS256")
    return {"success": True, "token": token}

async def get_all_blogs_admin():
    db = get_db()
    cursor = db["blogs"].find({}).sort("createdAt", -1)
    blogs = [serialize_blog(b) for b in await cursor.to_list(length=None)]
    return {"success": True, "blogs": blogs}

async def get_all_comments():
    db = get_db()
    pipeline = [
        {"$sort": {"createdAt": -1}},
        {"$lookup": {"from": "blogs", "localField": "blog", "foreignField": "_id", "as": "blogDoc"}},
        {"$unwind": {"path": "$blogDoc", "preserveNullAndEmptyArrays": True}}
    ]
    cursor = db["comments"].aggregate(pipeline)
    comments = []
    async for c in cursor:
        comments.append(serialize_comment(c))
    return {"success": True, "comments": comments}

async def get_dashboard():
    db = get_db()
    recent_cursor = db["blogs"].find({}).sort("createdAt", -1).limit(5)
    recent = [serialize_blog(b) for b in await recent_cursor.to_list(length=None)]
    blogs = await db["blogs"].count_documents({})
    comments = await db["comments"].count_documents({})
    drafts = await db["blogs"].count_documents({"isPublished": False})
    dashboardData = {"blogs": blogs, "comments": comments, "drafts": drafts, "recentBlogs": recent}
    return {"success": True, "dashboardData": dashboardData}

async def delete_comment_by_id(comment_id: str):
    db = get_db()
    await db["comments"].delete_one({"_id": ObjectId(comment_id)})
    return {"success": True, "message": "Comment deleted successfully"}

async def approve_comment_by_id(comment_id: str):
    db = get_db()
    await db["comments"].update_one({"_id": ObjectId(comment_id)}, {"$set": {"isApproved": True, "updatedAt": datetime.utcnow().isoformat()}})
    return {"success": True, "message": "Comment approved successfully"}

File: server/controllers/adminController.js
import jwt from 'jsonwebtoken'
import Blog from '../models/Blog.js';
import Comment from '../models/Comment.js';
import Subscriber from '../models/Subscriber.js';

export const adminLogin = async (req, res)=>{
    try {
        const {email, password} = req.body;

        const adminEmail = process.env.ADMIN_EMAIL || "admin@example.com";
        const adminPassword = process.env.ADMIN_PASSWORD || "admin123";
        const jwtSecret = process.env.JWT_SECRET || "your_jwt_secret_key_here";

                console.log("========== LOGIN DEBUG ==========");
        console.log("Entered Email:", email);
        console.log("Entered Password:", password);
        console.log("Admin Email:", adminEmail);
        console.log("Admin Password:", adminPassword);
        console.log("=================================");
        if(email !== adminEmail || password !== adminPassword){
            return res.json({success: false, message: "Invalid Credentials"})
        }

        const token = jwt.sign({email}, jwtSecret)
        res.json({success: true, token})
    } catch (error) {
        res.json({success: false, message: error.message})
    }
}

export const getAllBlogsAdmin = async (req, res) =>{
    try {
        const blogs = await Blog.find({}).sort({createdAt: -1});
        res.json({success: true, blogs})
    } catch (error) {
        res.json({success: false, message: error.message})
    }
}

export const getAllComments = async (req, res) =>{
    try {
        const comments = await Comment.find({}).populate("blog").sort({createdAt: -1})
        res.json({success: true, comments})
    } catch (error) {
        res.json({success: false, message: error.message})
    }
}

export const getDashboard = async (req, res) =>{
    try {
        const recentBlogs = await Blog.find({}).sort({ createdAt: -1 }).limit(5);
        const blogs = await Blog.countDocuments();
        const comments = await Comment.countDocuments()
        const drafts = await Blog.countDocuments({isPublished: false})

        const dashboardData = {
            blogs, comments, drafts, recentBlogs
        }
        res.json({success: true, dashboardData})
    } catch (error) {
        res.json({success: false, message: error.message})
    }
}

export const deleteCommentById = async (req, res) =>{
    try {
        const {id} = req.body;
        await Comment.findByIdAndDelete(id);
        res.json({success: true, message:"Comment deleted successfully" })
    } catch (error) {
       res.json({success: false, message: error.message}) 
    }
}

export const approveCommentById = async (req, res) =>{
    try {
        const {id} = req.body;
        await Comment.findByIdAndUpdate(id, {isApproved: true});
        res.json({success: true, message:"Comment approved successfully" })
    } catch (error) {
       res.json({success: false, message: error.message}) 
    }
}

export const getAnalytics = async (req, res) => {
    try {
        const blogs = await Blog.find({});
        const totalSubscribers = await Subscriber.countDocuments();
        
        let totalViews = 0;
        const categoryStats = {};

        blogs.forEach(b => {
            totalViews += b.views || 0;
            const cat = b.category || 'Uncategorized';
            if (!categoryStats[cat]) {
                categoryStats[cat] = { count: 0, views: 0 };
            }
            categoryStats[cat].count += 1;
            categoryStats[cat].views += b.views || 0;
        });

        const popularCategories = Object.keys(categoryStats).map(cat => ({
            category: cat,
            count: categoryStats[cat].count,
            views: categoryStats[cat].views
        })).sort((a, b) => b.views - a.views);

        const mostViewedPosts = [...blogs]
            .sort((a, b) => (b.views || 0) - (a.views || 0))
            .slice(0, 5);

        let recentSubscribers = [];
        try {
            recentSubscribers = await Subscriber.find({}).sort({ createdAt: -1 }).limit(5);
        } catch (e) {
            // ignore
        }

        res.json({
            success: true,
            analytics: {
                totalViews,
                totalBlogs: blogs.length,
                popularCategories,
                mostViewedPosts,
                totalSubscribers,
                recentSubscribers
            }
        });
    } catch (error) {
        res.json({ success: false, message: error.message });
    }
}